src/orion/client/runner.py

```diff
@@ -0,0 +1,323 @@
+# -*- coding: utf-8 -*-
+# pylint:disable=too-many-arguments
+# pylint:disable=too-many-instance-attributes
+"""
+Runner
+======
+
+Executes the optimization process
+"""
+import logging
+import time
+from contextlib import contextmanager
+from dataclasses import dataclass
+
+import orion.core
+from orion.core.utils.exceptions import (
+    BrokenExperiment,
+    CompletedExperiment,
+    InvalidResult,
+    ReservationRaceCondition,
+    ReservationTimeout,
+    WaitingForTrials,
+)
+from orion.core.utils.flatten import flatten, unflatten
+from orion.core.worker.consumer import ExecutionError
+from orion.core.worker.trial import AlreadyReleased
+from orion.executor.base import AsyncException, AsyncResult
+
+
+class LazyWorkers(Exception):
+    """Raised when all the workers have been idle for a given amount of time"""
+
+    pass
+
+
+def _optimize(trial, fct, trial_arg, **kwargs):
+    """Execute a trial on a worker"""
+    kwargs.update(flatten(trial.params))
+
+    if trial_arg:
+        kwargs[trial_arg] = trial
+
+    return fct(**unflatten(kwargs))
+
+
+log = logging.getLogger(__name__)
+
+
+@contextmanager
+def _timer(self, name):
+    start = time.time()
+    yield
+    total = time.time() - start
+
+    value = getattr(self, name)
+    setattr(self, name, value + total)
+
+
+@dataclass
+class _Stat:
+    sample: int = 0
+    scatter: int = 0
+    gather: int = 0
+
+    def time(self, name):
+        """Measure elapsed  time of a given block"""
+        return _timer(self, name)
+
+    def report(self):
+        """Show the elapsed time of different blocks"""
+        lines = [
+            f"Sample  {self.sample:7.4f}",
+            f"Scatter {self.scatter:7.4f}",
+            f"Gather  {self.gather:7.4f}",
+        ]
+        return "\n".join(lines)
+
+
+class Runner:
```

Comment:
Majority of the tests of ExperimentClient.workon are now tests for Runner. They should be moved to another test module for Runner specifically where additional tests will be needed for the new parts of Runner.

Reply:
most of those are using workon to access runner, I suggest we create a story to address this tech debt later

Reply:
We would be missing the tests for runner later on anyway. Moving the existing tests should be relatively painless.

Reply:
moved